src/System.Device.Gpio.Tests/GpioControllerTestBase.cs

```diff
@@ -130,30 +131,203 @@ public void OpenPinDefaultsModeToInput()
-                Thread.Sleep(1000);
-            Debugger.Break();
-
-            ManualResetEvent mre = new ManualResetEvent(false);
             bool wasCalled = false;
             using (GpioController controller = new GpioController(GetTestNumberingScheme(), GetTestDriver()))
             {
                 controller.OpenPin(InputPin, PinMode.Input);
                 controller.OpenPin(OutputPin, PinMode.Output);
                 controller.RegisterCallbackForPinValueChangedEvent(InputPin, PinEventTypes.Rising, callback);
                 controller.Write(OutputPin, PinValue.High);
-                mre.WaitOne(TimeSpan.FromSeconds(5));
+                Thread.Sleep(10);
                 Assert.True(wasCalled);
             }
 
             void callback(object sender, PinValueChangedEventArgs pinValueChangedEventArgs)
             {
                 wasCalled = true;
-                mre.Set();
+            }
+        }
+
+        [Fact]
+        public void AddCallbackFallingEdgeNotDetectedTest()
+        {
+            bool wasCalled = false;
+            using (GpioController controller = new GpioController(GetTestNumberingScheme(), GetTestDriver()))
+            {
+                controller.OpenPin(InputPin, PinMode.Input);
+                controller.OpenPin(OutputPin, PinMode.Output);
+                controller.Write(OutputPin, PinValue.Low);
+                controller.RegisterCallbackForPinValueChangedEvent(InputPin, PinEventTypes.Falling, callback);
+                controller.Write(OutputPin, PinValue.High);
+                Thread.Sleep(10);
+                Assert.False(wasCalled);
+            }
+
+            void callback(object sender, PinValueChangedEventArgs pinValueChangedEventArgs)
+            {
+                wasCalled = true;
+            }
+        }
+
+        [Fact]
+        public void AddCallbackRemoveCallbackTest()
+        {
+            int risingEventOccuredCount = 0, fallingEventOccuredCount = 0;
+            using (GpioController controller = new GpioController(GetTestNumberingScheme(), GetTestDriver()))
+            {
+                controller.OpenPin(InputPin, PinMode.Input);
+                controller.OpenPin(OutputPin, PinMode.Output);
+                controller.Write(OutputPin, PinValue.Low);
+
+                controller.RegisterCallbackForPinValueChangedEvent(InputPin, PinEventTypes.Rising, (o, e) => {
+                    risingEventOccuredCount++;
+                });                        
+                controller.RegisterCallbackForPinValueChangedEvent(InputPin, PinEventTypes.Rising, callback);
+                controller.RegisterCallbackForPinValueChangedEvent(InputPin, PinEventTypes.Rising, (o, e) => {
+                    risingEventOccuredCount++;
+                    if (fallingEventOccuredCount == 4)
+                    {
+                        controller.UnregisterCallbackForPinValueChangedEvent(InputPin, callback);
+                    }
```

Comment:
should there be an else-case here that causes the test to fail?

Reply:
No, that is not test failing scenario, its just random value for removing one of the event listeners in the middle of execution